### PR TITLE
Implement gdb watchpoints and writing

### PR DESCRIPTION
**Pull request by Arusekk**
Proof of concept usage in examples/console:

```c
>>> watch cons
Hardware watchpoint 1: cons
>>> c
Continuing.
─── Output/messages ────────────────────────────────────────────
Hardware watchpoint 1: cons

Old value = {
  bitmap = 0x0,
  font = 0x0,
  width = 0x0,
  height = 0x0,
  cursor = {
    x = 0x0,
    y = 0x0
  }
}
New value = {
  bitmap = 0xc7f828 <screen_bm>,
  font = 0x0,
  width = 0x0,
  height = 0x0,
  cursor = {
    x = 0x0,
    y = 0x0
  }
}
─── (...) ──────────────────────────────────────────────────────
>>>
```

Also implemented basic register writing and memory writing.

### CODE REVIEW

File: tools/debug/uae.py [hunk 1]
```diff
@@ -184,15 +186,28 @@ async def remove_breakpoint(self, addr):
...
-    async def remove_watchpoint(self, addr):
-        # {f <address>} Add/remove breakpoint.
-        lines = await self.communicate('f %X' % addr)
-        assert lines and lines[0] == 'Breakpoint removed'
+    async def insert_watchpoint(self, addr, size, kind='I'):
+        # w <num> <address> <length> <R/W/I/F/C> [<value>[.x]] (read/write/opcode/freeze/mustchange).
+        #                    Add/remove memory watchpoints.
+        key = (addr,size,kind)
+        index = self.watchpoints_r.index(None)
+        self.watchpoints[key] = index
```
Comment (cahirwpz) [inline tools/debug/uae.py:194]: Does it have to be that complicated given we'll have no more than 10 watchpoints (in 99% use cases)?

File: tools/debug/uae.py [hunk 2]
```diff
@@ -184,15 +186,28 @@ async def remove_breakpoint(self, addr):
...
+        self.watchpoints_r[index] = key
+        if index == len(self.watchpoints_r):
+          self.watchpoints_r.append(None)
+        lines = await self.communicate('w %d %X %d %s' % (index, addr, size, kind))
+        print(repr(lines))
```
Comment (cahirwpz) [inline tools/debug/uae.py:199]: Debugging leftover?

File: tools/debug/uae.py [hunk 3]
```diff
@@ -184,15 +186,28 @@ async def remove_breakpoint(self, addr):
...
+        assert lines and lines[-1] == 'Memwatch %d added' % index
+
+    async def remove_watchpoint(self, addr, size, kind='I'):
+        # w <num> <address> <length> <R/W/I/F/C> [<value>[.x]] (read/write/opcode/freeze/mustchange).
+        #                    Add/remove memory watchpoints.
+        key = (addr,size,kind)
+        index = self.watchpoints[key]
+        lines = await self.communicate('w %d' % index)
+        print(repr(lines))
```
Comment (cahirwpz) [inline tools/debug/uae.py:208]: Ditto!

File: tools/debug/uae.py [hunk 4]
```diff
@@ -184,15 +186,28 @@ async def remove_breakpoint(self, addr):
...
-        assert lines and lines[0] == 'Breakpoint added'
-
-    async def remove_watchpoint(self, addr):
-        # {f <address>} Add/remove breakpoint.
-        lines = await self.communicate('f %X' % addr)
-        assert lines and lines[0] == 'Breakpoint removed'
+    async def insert_watchpoint(self, addr, size, kind='I'):
+        # w <num> <address> <length> <R/W/I/F/C> [<value>[.x]] (read/write/opcode/freeze/mustchange).
+        #                    Add/remove memory watchpoints.
+        key = (addr,size,kind)
```
Comment (cahirwpz) [inline tools/debug/uae.py:192]: Have you considered using `namedtuple` ?

File: tools/debug/uae.py [hunk 5]
```diff
@@ -206,6 +221,11 @@ async def entry_point(self):
     async def prologue(self):
         lines = await self.recv()
         data = {}
+        # Memwatch 2: break at 00C7FDC0.W  W  0000000F PC=00C7BFEC CPUDW (000)
+        if lines[0].startswith('Memwatch'):
+            line = lines.pop(0)
+            data['watch'] = int(line.split()[4].split('.')[0], 16)
+            data['cause'] = 'watch:%x' % data['watch']
```
Comment (cahirwpz) [inline tools/debug/uae.py:228]: Would it make sense to add `data['cause']` also for _breakpoints_ and _exceptions_ ?

File: tools/debug/gdb.py [hunk 6]
```diff
@@ -226,8 +240,9 @@ async def handle_packet(self, packet):
             self.gdb.send_ack('DEAD{:04X}'.format(int(packet[1:], 16)))
         elif packet[0] == 'P':
             # Write register n… with value r….
-            # TODO {P n…=r…}
-            self.gdb.send_ack('E00')
+            reg, val = map(lambda x: int(x, 16), packet[1:].split('='))
```
[the remark this answers is not in the document]
Comment (Arusekk) [reply tools/debug/gdb.py:243]: I was just up to, but then I saw this pattern all around the file.

File: tools/debug/uae.py (hunk 5 above)
Comment (Arusekk) [reply tools/debug/uae.py:228]: Citing https://sourceware.org/gdb/onlinedocs/gdb/Stop-Reply-Packets.html#Stop-Reply-Packets
> If n is a recognized stop reason, it describes a more specific event that stopped the target. The currently defined stop reasons are listed below. The aa should be ‘05’, the trap signal. At most one stop reason should be present.

There is also a remark about other causes that they should be sent only if gdb indicates that they are `qSupported`. So I just did not add them.

File: tools/debug/gdb.py [hunk 7]
```diff
@@ -184,6 +184,20 @@ async def handle_query(self, packet):
 
         return True
 
+    @staticmethod
+    def binary_decode(data):
```
[the remark this answers is not in the document]
Comment (Arusekk) [reply tools/debug/gdb.py:188]: This is special GDB's protocol for escaping binary data transfer: https://sourceware.org/gdb/onlinedocs/gdb/Overview.html#Binary-Data

File: tools/debug/uae.py (hunk 4 above)
Comment (Arusekk) [reply tools/debug/uae.py]: I think `namedtuple` would be an overkill — this triple is used just in these two functions, neighboring in code.

File: tools/debug/gdb.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cahirwpz) [reply tools/debug/gdb.py:245]: Ok.

**Pull request merged**